File: windows/DisplayShot/Overlay/SelectionModel.cs
using System.Windows;

namespace DisplayShot.Overlay;

public enum Handle { TopLeft, Top, TopRight, Right, BottomRight, Bottom, BottomLeft, Left }

public static class HandleExtensions
{
    public static bool AffectsLeft(this Handle h) => h is Handle.TopLeft or Handle.Left or Handle.BottomLeft;
    public static bool AffectsRight(this Handle h) => h is Handle.TopRight or Handle.Right or Handle.BottomRight;
    public static bool AffectsTop(this Handle h) => h is Handle.TopLeft or Handle.Top or Handle.TopRight;
    public static bool AffectsBottom(this Handle h) => h is Handle.BottomLeft or Handle.Bottom or Handle.BottomRight;
    public static bool IsCorner(this Handle h) => (h.AffectsLeft() || h.AffectsRight()) && (h.AffectsTop() || h.AffectsBottom());

    public static Handle? Make(bool left, bool right, bool top, bool bottom) => (left, right, top, bottom) switch
    {
        (true, false, true, false) => Handle.TopLeft,
        (false, false, true, false) => Handle.Top,
        (false, true, true, false) => Handle.TopRight,
        (false, true, false, false) => Handle.Right,
        (false, true, false, true) => Handle.BottomRight,
        (false, false, false, true) => Handle.Bottom,
        (true, false, false, true) => Handle.BottomLeft,
        (true, false, false, false) => Handle.Left,
        _ => null,
    };

    public static Handle FlippedHorizontally(this Handle h) =>
        Make(h.AffectsRight(), h.AffectsLeft(), h.AffectsTop(), h.AffectsBottom()) ?? h;

    public static Handle FlippedVertically(this Handle h) =>
        Make(h.AffectsLeft(), h.AffectsRight(), h.AffectsBottom(), h.AffectsTop()) ?? h;
}

public abstract record SelectionHit
{
    public sealed record HandleHit(Handle Handle) : SelectionHit;
    public sealed record Inside : SelectionHit;
    public sealed record Outside : SelectionHit;
}

/// <summary>The selection rectangle in canvas pixels, always kept inside <see cref="Bounds"/>.</summary>
public sealed class SelectionModel
{
    public const double MinSize = 1;
    public const double HandleSize = 8;
    public const double HandleHitRadius = 9;

    public Rect Rect { get; private set; }
    public Rect Bounds { get; }

    public SelectionModel(Rect rect, Rect bounds)
    {
        Bounds = bounds;
        Rect = Normalize(rect, bounds);
    }

    public static double Clamp(double v, double lo, double hi) => Math.Min(Math.Max(v, lo), hi);

    public static Point ClampPoint(Point p, Rect r) => new(Clamp(p.X, r.Left, r.Right), Clamp(p.Y, r.Top, r.Bottom));

    /// <summary>Shifts the rect into bounds, shrinking only if it is larger than bounds.</summary>
    public static Rect Fit(Rect r, Rect bounds)
    {
        var w = Math.Min(r.Width, bounds.Width);
        var h = Math.Min(r.Height, bounds.Height);
        var x = r.X;
        var y = r.Y;
        if (x < bounds.Left) x = bounds.Left;
        if (x + w > bounds.Right) x = bounds.Right - w;
        if (y < bounds.Top) y = bounds.Top;
        if (y + h > bounds.Bottom) y = bounds.Bottom - h;
        return new Rect(x, y, w, h);
    }

    public static Rect Normalize(Rect r, Rect bounds)
    {
        var x0 = Clamp(r.Left, bounds.Left, bounds.Right);
        var x1 = Clamp(r.Right, bounds.Left, bounds.Right);
        var y0 = Clamp(r.Top, bounds.Top, bounds.Bottom);
        var y1 = Clamp(r.Bottom, bounds.Top, bounds.Bottom);
        return Fit(new Rect(x0, y0, Math.Max(MinSize, x1 - x0), Math.Max(MinSize, y1 - y0)), bounds);
    }

    /// <summary>Rect for a fresh drag; <paramref name="square"/> is the Shift constraint.</summary>
    public static Rect RectFromDrag(Point origin, Point current, bool square, Rect bounds)
    {
        var c = ClampPoint(current, bounds);
        var dx = c.X - origin.X;
        var dy = c.Y - origin.Y;
        if (square)
        {
            var side = Math.Max(Math.Abs(dx), Math.Abs(dy));
            var limX = dx >= 0 ? bounds.Right - origin.X : origin.X - bounds.Left;
            var limY = dy >= 0 ? bounds.Bottom - origin.Y : origin.Y - bounds.Top;
            side = Math.Min(side, Math.Min(limX, limY));
            dx = (dx < 0 ? -1 : 1) * side;
            dy = (dy < 0 ? -1 : 1) * side;
        }
        return new Rect(origin, new Point(origin.X + dx, origin.Y + dy));
    }

    public Point HandleCenter(Handle h)
    {
        var x = h.AffectsLeft() ? Rect.Left : h.AffectsRight() ? Rect.Right : Rect.Left + Rect.Width / 2;
        var y = h.AffectsTop() ? Rect.Top : h.AffectsBottom() ? Rect.Bottom : Rect.Top + Rect.Height / 2;
        return new Point(x, y);
    }

    public IEnumerable<(Handle Handle, Rect Rect)> HandleRects()
    {
        foreach (var h in Enum.GetValues<Handle>())
        {
            var c = HandleCenter(h);
            yield return (h, new Rect(c.X - HandleSize / 2, c.Y - HandleSize / 2, HandleSize, HandleSize));
        }
    }

    public SelectionHit HitTest(Point p)
    {
        var handles = Enum.GetValues<Handle>();
        foreach (var h in handles.Where(h => h.IsCorner()).Concat(handles.Where(h => !h.IsCorner())))
        {
            if ((HandleCenter(h) - p).Length <= HandleHitRadius) return new SelectionHit.HandleHit(h);
        }
        var inflated = Rect;
        inflated.Inflate(2, 2);
        return inflated.Contains(p) ? new SelectionHit.Inside() : new SelectionHit.Outside();
    }

    /// <summary>Drags a handle. Returns the handle now under the pointer (flips when inverted).</summary>
    public Handle Resize(Handle handle, Point point, bool square)
    {
        var p = ClampPoint(point, Bounds);
        double l = Rect.Left, r = Rect.Right, t = Rect.Top, b = Rect.Bottom;

        if (square && handle.IsCorner())
        {
            var ax = handle.AffectsLeft() ? r : l;
            var ay = handle.AffectsTop() ? b : t;
            var sx = p.X >= ax ? 1 : -1;
            var sy = p.Y >= ay ? 1 : -1;
            var side = Math.Max(Math.Abs(p.X - ax), Math.Abs(p.Y - ay));
            var limX = sx > 0 ? Bounds.Right - ax : ax - Bounds.Left;
            var limY = sy > 0 ? Bounds.Bottom - ay : ay - Bounds.Top;
            side = Math.Min(side, Math.Min(limX, limY));
            var nx = ax + sx * side;
            var ny = ay + sy * side;
            Rect = Normalize(new Rect(new Point(ax, ay), new Point(nx, ny)), Bounds);
            return HandleExtensions.Make(sx < 0, sx > 0, sy < 0, sy > 0) ?? handle;
        }

        if (handle.AffectsLeft()) l = p.X;
        if (handle.AffectsRight()) r = p.X;
        if (handle.AffectsTop()) t = p.Y;
        if (handle.AffectsBottom()) b = p.Y;
        var h = handle;
        if (l > r) { (l, r) = (r, l); h = h.FlippedHorizontally(); }
        if (t > b) { (t, b) = (b, t); h = h.FlippedVertically(); }
        Rect = Normalize(new Rect(l, t, r - l, b - t), Bounds);
        return h;
    }

    public void MoveTo(Point origin) => Rect = Fit(new Rect(origin, Rect.Size), Bounds);

    public void Nudge(double dx, double dy) => MoveTo(new Point(Rect.X + dx, Rect.Y + dy));

    /// <summary>Grows/shrinks the right and bottom edges.</summary>
    public void Grow(double dw, double dh)
    {
        var w = Clamp(Rect.Width + dw, MinSize, Bounds.Right - Rect.Left);
        var h = Clamp(Rect.Height + dh, MinSize, Bounds.Bottom - Rect.Top);
        Rect = new Rect(Rect.X, Rect.Y, w, h);
    }
}

File: macos/Sources/DisplayShotKit/Overlay/SelectionModel.swift
import CoreGraphics

enum Handle: CaseIterable, Equatable {
    case topLeft, top, topRight, right, bottomRight, bottom, bottomLeft, left

    var affectsLeft: Bool { self == .topLeft || self == .left || self == .bottomLeft }
    var affectsRight: Bool { self == .topRight || self == .right || self == .bottomRight }
    var affectsTop: Bool { self == .topLeft || self == .top || self == .topRight }
    var affectsBottom: Bool { self == .bottomLeft || self == .bottom || self == .bottomRight }
    var isCorner: Bool { (affectsLeft || affectsRight) && (affectsTop || affectsBottom) }

    static func make(left: Bool, right: Bool, top: Bool, bottom: Bool) -> Handle? {
        switch (left, right, top, bottom) {
        case (true, false, true, false): return .topLeft
        case (false, false, true, false): return .top
        case (false, true, true, false): return .topRight
        case (false, true, false, false): return .right
        case (false, true, false, true): return .bottomRight
        case (false, false, false, true): return .bottom
        case (true, false, false, true): return .bottomLeft
        case (true, false, false, false): return .left
        default: return nil
        }
    }

    var flippedHorizontally: Handle {
        Handle.make(left: affectsRight, right: affectsLeft, top: affectsTop, bottom: affectsBottom) ?? self
    }

    var flippedVertically: Handle {
        Handle.make(left: affectsLeft, right: affectsRight, top: affectsBottom, bottom: affectsTop) ?? self
    }
}

enum SelectionHit: Equatable {
    case handle(Handle)
    case inside
    case outside
}

/// The selection rectangle on one display, in points with a top-left origin, always kept
/// inside `bounds`.
struct SelectionModel: Equatable {
    static let minSize: CGFloat = 1
    static let handleSize: CGFloat = 8
    static let handleHitRadius: CGFloat = 9

    private(set) var rect: CGRect
    let bounds: CGRect

    init(rect: CGRect, bounds: CGRect) {
        self.bounds = bounds
        self.rect = SelectionModel.normalize(rect, in: bounds)
    }

    static func normalize(_ r: CGRect, in bounds: CGRect) -> CGRect {
        let s = r.standardized
        let x0 = clamp(s.minX, bounds.minX, bounds.maxX), x1 = clamp(s.maxX, bounds.minX, bounds.maxX)
        let y0 = clamp(s.minY, bounds.minY, bounds.maxY), y1 = clamp(s.maxY, bounds.minY, bounds.maxY)
        let out = CGRect(x: x0, y: y0, width: max(minSize, x1 - x0), height: max(minSize, y1 - y0))
        return out.fitted(in: bounds)
    }

    /// Rect for a fresh drag from `origin` to `current`; `square` is the Shift constraint.
    static func rect(from origin: CGPoint, to current: CGPoint, square: Bool, in bounds: CGRect) -> CGRect {
        let c = current.clamped(to: bounds)
        var dx = c.x - origin.x, dy = c.y - origin.y
        if square {
            var side = max(abs(dx), abs(dy))
            let limX = dx >= 0 ? bounds.maxX - origin.x : origin.x - bounds.minX
            let limY = dy >= 0 ? bounds.maxY - origin.y : origin.y - bounds.minY
            side = min(side, limX, limY)
            dx = (dx < 0 ? -1 : 1) * side
            dy = (dy < 0 ? -1 : 1) * side
        }
        return CGRect(corner: origin, CGPoint(x: origin.x + dx, y: origin.y + dy))
    }

    func handleCenter(_ h: Handle) -> CGPoint {
        let x: CGFloat = h.affectsLeft ? rect.minX : (h.affectsRight ? rect.maxX : rect.midX)
        let y: CGFloat = h.affectsTop ? rect.minY : (h.affectsBottom ? rect.maxY : rect.midY)
        return CGPoint(x: x, y: y)
    }

    func handleRects() -> [(Handle, CGRect)] {
        let s = SelectionModel.handleSize
        return Handle.allCases.map { h in
            let c = handleCenter(h)
            return (h, CGRect(x: c.x - s / 2, y: c.y - s / 2, width: s, height: s))
        }
    }

    func hitTest(_ p: CGPoint) -> SelectionHit {
        let ordered = Handle.allCases.filter(\.isCorner) + Handle.allCases.filter { !$0.isCorner }
        for h in ordered where handleCenter(h).distance(to: p) <= SelectionModel.handleHitRadius {
            return .handle(h)
        }
        return rect.insetBy(dx: -2, dy: -2).contains(p) ? .inside : .outside
    }

    /// Drags `handle` to `point`. Returns the handle that now sits under the pointer (it flips
    /// when the rectangle is dragged through itself).
    @discardableResult
    mutating func resize(_ handle: Handle, to point: CGPoint, square: Bool) -> Handle {
        let p = point.clamped(to: bounds)
        var l = rect.minX, r = rect.maxX, t = rect.minY, b = rect.maxY

        if square && handle.isCorner {
            let ax = handle.affectsLeft ? r : l
            let ay = handle.affectsTop ? b : t
            let sx: CGFloat = p.x >= ax ? 1 : -1
            let sy: CGFloat = p.y >= ay ? 1 : -1
            var side = max(abs(p.x - ax), abs(p.y - ay))
            let limX = sx > 0 ? bounds.maxX - ax : ax - bounds.minX
            let limY = sy > 0 ? bounds.maxY - ay : ay - bounds.minY
            side = min(side, limX, limY)
            let nx = ax + sx * side, ny = ay + sy * side
            rect = SelectionModel.normalize(CGRect(corner: CGPoint(x: ax, y: ay), CGPoint(x: nx, y: ny)), in: bounds)
            return Handle.make(left: sx < 0, right: sx > 0, top: sy < 0, bottom: sy > 0) ?? handle
        }

        if handle.affectsLeft { l = p.x }
        if handle.affectsRight { r = p.x }
        if handle.affectsTop { t = p.y }
        if handle.affectsBottom { b = p.y }
        var h = handle
        if l > r { swap(&l, &r); h = h.flippedHorizontally }
        if t > b { swap(&t, &b); h = h.flippedVertically }
        rect = SelectionModel.normalize(CGRect(x: l, y: t, width: r - l, height: b - t), in: bounds)
        return h
    }

    mutating func move(to origin: CGPoint) {
        rect = CGRect(origin: origin, size: rect.size).fitted(in: bounds)
    }

    mutating func nudge(dx: CGFloat, dy: CGFloat) {
        move(to: CGPoint(x: rect.minX + dx, y: rect.minY + dy))
    }

    /// Grows/shrinks the right and bottom edges.
    mutating func grow(dw: CGFloat, dh: CGFloat) {
        var r = rect
        r.size.width = clamp(r.width + dw, SelectionModel.minSize, bounds.maxX - r.minX)
        r.size.height = clamp(r.height + dh, SelectionModel.minSize, bounds.maxY - r.minY)
        rect = r
    }
}
